### PR TITLE
feat / KuCoin connector

### PR DESCRIPTION
**Pull request by vic-en**
Initial work for KuCoin connector consisting majorly of the Data Source and Orderbook Tracker files along with Auth file towards completing the connector.

### CODE REVIEW

File: hummingbot/market/kucoin/kucoin_api_order_book_data_source.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+            market_data: List[Dict[str, Any]] = [
+                {**item, **trading_pairs[item["symbol"]]}
+                for item in market_data["data"]["ticker"]
+                if item["symbol"] in trading_pairs
+            ]
+
+            # Build the data frame.
+            all_markets: pd.DataFrame = pd.DataFrame.from_records(data=market_data, index="symbol")
+            all_markets.loc[:, "USDVolume"] = all_markets.volvalue
```
Comment (CrimsonJacket) [inline hummingbot/market/kucoin/kucoin_api_order_book_data_source.py]: The `volvalue` here does not necessarily mean the `USDVolume` for the particular trading pair. 
According to their API, it is just 

> The executed amount in quote currency

In cases like `LOOM-BTC`, you will be required to multiply the `volvalue` field with the `last` field(last trade price) of `BTC-USDT` to obtain the `USDVolume` for the trading pair.

File: hummingbot/market/kucoin/kucoin_api_order_book_data_source.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+                self.logger().network(
+                    f"Error getting active exchange information.",
+                    exc_info=True,
+                    app_warning_msg=f"Error getting active exchange information. Check network connection."
+                )
+        return self._symbols
+
+    @staticmethod
+    async def get_snapshot(client: aiohttp.ClientSession, trading_pair: str) -> Dict[str, Any]:
+        params: str = "?symbol=" + trading_pair
```
Comment (CrimsonJacket) [inline hummingbot/market/kucoin/kucoin_api_order_book_data_source.py]: Use `urlencode` instead to ensure proper encoding is done on url parameters in REST API calls

File: hummingbot/market/kucoin/kucoin_api_order_book_data_source.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+	#get required data to create a websocket request
+	async def ws_connect_data(self):
+		async with aiohttp.ClientSession() as session:
+			async with session.post('https://api.kucoin.com/api/v1/bullet-public', data=b'') as resp:
+				response: aiohttp.ClientResponse = resp
+				if response.status != 200:
+					raise IOError(f"Error fetching Kucoin websocket connection data."
+								  f"HTTP status is {response.status}.")
+				api_data = await response.read()
+				data: Dict[str, Any] = json.loads(api_data)
```
Comment (CrimsonJacket) [inline hummingbot/market/kucoin/kucoin_api_order_book_data_source.py]: Small matter,
Considering that KuCoin documentation did mention that >All requests and responses are application/json content type.

You can simply do the following:
`data: Dict[str, Any] = await response.json()`

**Comment by CrimsonJacket** [comment]
Do make sure to following the instructions as defined in [CONTRIBUTING.md](https://github.com/CoinAlpha/hummingbot/blob/master/CONTRIBUTING.md), when making a PR

File: hummingbot/market/kucoin/kucoin_order_book_tracker.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        message_queue: asyncio.Queue = self._tracking_message_queues[symbol]
+        order_book: OrderBook = self._order_books[symbol]
+        last_message_timestamp: float = time.time()
+        diff_messages_accepted: int = 0
+
+        while True:
+            try:
+                message: OrderBookMessage = await message_queue.get()
+                if message.type is OrderBookMessageType.DIFF:
+                    # Kucoin websocket messages contain the entire order book state so they should be treated as snapshots
```
Comment (chowryan) [inline hummingbot/market/kucoin/kucoin_order_book_tracker.py]: Seems like Kucoin has snapshots with diffs. Please use this instead

https://docs.kucoin.com/#level-2-market-data

Comment (vic-en) [reply hummingbot/market/kucoin/kucoin_order_book_tracker.py]: I already implemented the level2-market data for in both restful and websocket for snapshot and diff respectively. Are you requesting that I remove the `elif message.type is OrderBookMessageType.SNAPSHOT:` statement?

Comment (chowryan) [reply hummingbot/market/kucoin/kucoin_order_book_tracker.py]: according to their docs diffs are updates. if you call apply_snapshot doesnt it replace the whole orderbook with just  orders in the diff?

**Comment by vic-en** [comment]
I have updated code and included test. Kindly check again.

File: hummingbot/market/kucoin/kucoin_api_order_book_data_source.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+from hummingbot.core.data_type.order_book_tracker_data_source import OrderBookTrackerDataSource
+from hummingbot.core.data_type.order_book_tracker_entry import OrderBookTrackerEntry
+from hummingbot.core.data_type.order_book_message import OrderBookMessage
+from hummingbot.core.data_type.order_book import OrderBook
+from hummingbot.logger import HummingbotLogger
+from hummingbot.market.kucoin.kucoin_order_book import KucoinOrderBook
+
+TRADING_PAIR_FILTER = re.compile(r"(BTC|ETH|USDT)$")
+
+SNAPSHOT_REST_URL = "https://api.kucoin.com/api/v1/market/orderbook/level2_100"
```
Comment (CrimsonJacket) [inline hummingbot/market/kucoin/kucoin_api_order_book_data_source.py]: According to KuCoin's documentation, this endpoint only returns 
>  100 pieces of data (ask and bid data) on the order book 
I would recommend that you follow the instructions as stated in their documentation on maintaining a up-to-date Order Book. 

As stated [here](https://docs.kucoin.com/#level-2-market-data) they recommend that to maintain an order book, we use this `/api/v2/market/orderbook/level2` endpoint before applying the incremental market data from the websocket.

File: hummingbot/market/kucoin/kucoin_order_book_tracker.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+                    order_book.apply_snapshot(message.bids, message.asks, message.update_id)
```
Comment (chowryan) [inline hummingbot/market/kucoin/kucoin_order_book_tracker.py]: you are using `apply_snapshot` for diff messages. you need to be using apply_diffs for diffs
```
{
  "type":"message",
  "topic":"/market/level2:BTC-USDT",
  "subject":"trade.l2update",
  "data":{
    "sequenceStart":1545896669105,
    "sequenceEnd":1545896669106,
    "symbol":"BTC-USDT",
    "changes":{
      "asks":[["6","1","1545896669105"]],           //price, size, sequence
      "bids":[["4","1","1545896669106"]]
    }
  }
}
```

Comment (chowryan) [reply hummingbot/market/kucoin/kucoin_order_book_tracker.py]: see binance as an example
image

Comment (chowryan) [reply hummingbot/market/kucoin/kucoin_order_book_tracker.py]: nvm i was looking at outdated version

File: hummingbot/market/kucoin/kucoin_api_order_book_data_source.py [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+            if exchange_response.status != 200:
+                raise IOError(f"Error fetching Kucoin exchange information. "
+                              f"HTTP status is {exchange_response.status}.")
+
+            market_data = await market_response.json()
+            exchange_data = await exchange_response.json()
+            
+            attr_name_map = {"baseCurrency": "baseAsset", "quoteCurrency": "quoteAsset"}
+
+            trading_pairs: Dict[str, Any] = {item["symbol"]: {attr_name_map[k]: item[k] for k in ["baseAsset", "quoteAsset"]}
```
Comment (CrimsonJacket) [inline hummingbot/market/kucoin/kucoin_api_order_book_data_source.py]: I am getting a `KeyError` here.

I think you meant to write
> trading_pairs: Dict[str, Any] = {... for k in ["**baseCurrency**", "**quoteCurrency**"]}

File: hummingbot/market/kucoin/kucoin_api_order_book_data_source.py [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+                    mul: str = quote_currency + "-USDT"
+                    div: str = "USDT-" + quote_currency
+                    if all_markets.loc[mul]:
+                        all_markets.loc[:, "USDVolume"] = all_markets.volvalue * all_markets.loc[mul].last
+                    else:
+                        all_markets.loc[:, "USDVolume"] = all_markets.volvalue / all_markets.loc[div].last
+                except Exception:
+                    self.logger().error(f"Unable to get USD Volume for {trading_pair}. ", exc_info=True)
+            all_markets.loc[:, "volume"] = all_markets.quoteVolume
+            return all_markets.sort_values("USDVolume", ascending=False)
```
Comment (CrimsonJacket) [inline hummingbot/market/kucoin/kucoin_api_order_book_data_source.py]: I understand that this portion of code is meant to generate the `USDVolume` column for the `DataFrame` but there are definitely some stuff you need to fine tune. Considering that I am not getting any desirable output from the `get_active_exchange_markets()` function.

You might want to read up on the `pandas` library, in particular the documentation on [`DataFrame`](https://pandas.pydata.org/pandas-docs/stable/reference/frame.html).

Also you might want to refer to [Binance](https://github.com/CoinAlpha/hummingbot/blob/f8d1c19dfd0875bd12717f9c46ddbe20cc7b9a0d/hummingbot/market/binance/binance_api_order_book_data_source.py#L57), [Bittrex](https://github.com/CoinAlpha/hummingbot/blob/f8d1c19dfd0875bd12717f9c46ddbe20cc7b9a0d/hummingbot/market/bittrex/bittrex_api_order_book_data_source.py#L61) or [CoinbasePro](https://github.com/CoinAlpha/hummingbot/blob/f8d1c19dfd0875bd12717f9c46ddbe20cc7b9a0d/hummingbot/market/coinbase_pro/coinbase_pro_api_order_book_data_source.py#L56) to generate the relevant data.

Comment (vic-en) [reply hummingbot/market/kucoin/kucoin_api_order_book_data_source.py]: Solved. I've corrected the get_active_exchange_markets() to give correct USD Volume.

File: hummingbot/market/kucoin/kucoin_api_order_book_data_source.py (hunk 7 above)
Comment (vic-en) [reply hummingbot/market/kucoin/kucoin_api_order_book_data_source.py]: Corrected, thanks.

File: hummingbot/market/kucoin/kucoin_api_order_book_data_source.py (hunk 5 above)
Comment (vic-en) [reply hummingbot/market/kucoin/kucoin_api_order_book_data_source.py]: Changed SNAPSHOT_REST_URL to /api/v2/market/orderbook/level2. Confirm in https://github.com/vic-en/hummingbot/blob/feat/kucoin-connector/hummingbot/market/kucoin/kucoin_api_order_book_data_source.py

File: hummingbot/market/kucoin/kucoin_api_order_book_data_source.py [hunk 9]
```diff
@@ -91,23 +91,21 @@ def __init__(self, symbols: Optional[List[str]] = None):
...
+                    all_markets.loc[product_name, "USDVolume"] = float(row.volValue)
+                else:
+                        quote_currency: str = product_name.split('-')[1]
+                        mul: str = quote_currency + "-USDT"
+                        div: str = "USDT-" + quote_currency
+                        if mul in all_markets.index:
+                            all_markets.loc[product_name, "USDVolume"] = float(row.volValue) * float(all_markets.loc[mul, "last"])
+                        elif div in all_markets.index:
+                            all_markets.loc[product_name, "USDVolume"] = float(row.volValue) / float(all_markets.loc[div, "last"])
+            all_markets.loc[:, "volume"] = all_markets.vol
```
Comment (CrimsonJacket) [inline hummingbot/market/kucoin/kucoin_api_order_book_data_source.py]: `get_active_exchange_markets()` is working now 👍 
However, it seems that it takes pretty long to derive the result. I would recommend modifying this sometime later. Maybe I will write something out myself when I have the chance

Comment (vic-en) [reply hummingbot/market/kucoin/kucoin_api_order_book_data_source.py]: Alright, I will. Thanks.

Comment (spmuppar) [reply hummingbot/market/kucoin/kucoin_api_order_book_data_source.py]: @vic-en Can I get in touch with you regarding exchange connectors. ?

Comment (vic-en) [reply hummingbot/market/kucoin/kucoin_api_order_book_data_source.py]: @spmuppar, you can

Comment (spmuppar) [reply hummingbot/market/kucoin/kucoin_api_order_book_data_source.py]: @vic-en what's your email id or discord id ?.
If you don't want share those details, please ping me on praneeth@pec.edu or my discord id "spmuppar#3706"

File: test/integration/test_kucoin_user_stream_tracker.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+# logging.basicConfig(level=logging.DEBUG)
+
+
+class KucoinOrderBookTrackerUnitTest(unittest.TestCase):
+    order_book_tracker: Optional[KucoinOrderBookTracker] = None
+    @classmethod
+    def setUpClass(cls):
+        cls.ev_loop: asyncio.BaseEventLoop = asyncio.get_event_loop()
+        cls.kucoin_auth = KucoinAuth(conf.kucoin_api_key, conf.kucoin_secret_key, conf.kucoin_passphrase)
```
Comment (CrimsonJacket) [inline test/integration/test_kucoin_user_stream_tracker.py]: According to `KucoinAuth` the arguments should be as 
> KucoinAuth(api_key, passphrase, secret_key)

**Pull request closed without merging**